services/npm-stat/npm-stat-downloads.service.js

```diff
@@ -0,0 +1,83 @@
+import Joi from 'joi'
+import { nonNegativeInteger } from '../validators.js'
+import { BaseJsonService } from '../index.js'
+import { renderDownloadsBadge } from '../downloads.js'
+
+const schema = Joi.object()
+  .pattern(Joi.string(), Joi.object().pattern(Joi.string(), nonNegativeInteger))
+  .required()
+
+const NPM_INITIAL_RELEASE_DATE = '2010-01-12'
+
+const MILLISECONDS_IN_A_DAY = 1000 * 60 * 60 * 24
+const MILLISECONDS_IN_A_WEEK = MILLISECONDS_IN_A_DAY * 7
+const MILLISECONDS_IN_A_MONTH = MILLISECONDS_IN_A_DAY * 30
+const MILLISECONDS_IN_A_YEAR = MILLISECONDS_IN_A_DAY * 365
+
+const intervalMap = {
+  dw: { interval: 'week', difference: MILLISECONDS_IN_A_WEEK },
+  dm: { interval: 'month', difference: MILLISECONDS_IN_A_MONTH },
+  dy: { interval: 'year', difference: MILLISECONDS_IN_A_YEAR },
+  dt: { difference: -1 },
+}
+
+export default class NpmStatDownloads extends BaseJsonService {
+  static category = 'downloads'
+
+  static route = {
+    base: 'npm-stat',
+    pattern: ':interval(dw|dm|dy|dt)/:author',
+  }
+
+  static examples = [
+    {
+      title: 'npm (by author)',
+      documentation:
+        'The total number of downloads of npm packages published by the specified author.',
+      namedParams: { interval: 'dy', author: 'dukeluo' },
+      staticPreview: this.render({ interval: 'dy', downloadCount: 30000 }),
+      keywords: ['node'],
+    },
+  ]
+
+  static _cacheLength = 21600
+
+  static defaultBadgeData = { label: 'downloads' }
+
+  static getTotalDownloads(data) {
```

Comment:
I think it would be useful to add a unit test for this function, just to nail down that it is correctly counting in the case where the user has multiple packages. A mistake here wouldn't break the service tests. Could we add a `.spec` file with a test or two for the `getTotalDownloads()` function?

Reply:
Yes, I agree. I will do this.